### PR TITLE
Add new AWS Foundational Security Best Practices controls. closes #258

### LINKED ISSUE #258: Add new AWS Foundational Security Best Practices controls.
Adding new FSBP controls

1. APIGateway.5
2. EC2.15
3. EC2.19
4. ElasticBeanstalk.1
5. ELB.7
6. Lambda.4
7. RDS.18
8. RDS.19
9. RDS.20
10. RDS.21
11. RDS.22
12. RDS.23
13. SQS.1

### PR DESCRIPTION
**Pull request by misraved**
### Checklist
- [ ] Issue(s) linked

### CODE REVIEW

File: query/ec2/ec2_classic_loadbalancer_connection_draining_enabled.sql [hunk 1]
```diff
@@ -0,0 +1,16 @@
+select
```
Comment (khushboo9024) [inline query/ec2/ec2_classic_loadbalancer_connection_draining_enabled.sql:1]: ec2_classic_loadbalancer_connection_draining_enabled.sql ->  ec2_classic_lb_connection_draining_enabled.sql

File: query/ec2/ec2_classic_loadbalancer_connection_draining_enabled.sql [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  -- Required Columns
+  arn as resource,
+  case
+    when connection_draining_enabled = 'true' then 'ok'
```
Comment (khushboo9024) [inline query/ec2/ec2_classic_loadbalancer_connection_draining_enabled.sql:5]:
```suggestion
    when connection_draining_enabled then 'ok'
```

File: query/rds/rds_db_instance_and_cluster_no_default_port.sql [hunk 3]
```diff
@@ -0,0 +1,49 @@
+(
```
Comment (khushboo9024) [inline query/rds/rds_db_instance_and_cluster_no_default_port.sql:1]: rds_db_instance_and_cluster_no_default_port.sql -> rds_db_instance_and_cluster_uses_default_port.sql

File: query/sqs/sqs_queue_encrypted_at_rest.sql [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  when kms_master_key_id is null then 'alarm'
+  else 'ok'
+end as status,
+case
+  when kms_master_key_id is null then title || ' encryption at rest disabled.'
+  else title || ' encryption at rest enabled.'
+end as reason,
+  -- Additional Dimensions
+  region,
+  account_id
```
Comment (khushboo9024) [inline query/sqs/sqs_queue_encrypted_at_rest.sql:14]: Indentation is wrong.

File: query/vpc/vpc_security_group_restricted_common_ports.sql [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+)
+select
+  -- Required Columns
+  arn as resource,
+  case
+    when ingress_ssh_rules.group_id is null then 'ok'
+    else 'alarm'
+  end as status,
+  case
+    when ingress_ssh_rules.group_id is null then sg.group_id || ' ingress restricted for ports 20, 21, 22, 3306, 3389, 4333 from 0.0.0.0/0.'
```
Comment (khushboo9024) [inline query/vpc/vpc_security_group_restricted_common_ports.sql:102]: Reason needs modification, we are not just restricting 20, 21, 22, 3306, 3389, 4333 ports.

File: query/vpc/vpc_subnet_auto_assign_public_ip_disabled.sql [hunk 6]
```diff
@@ -0,0 +1,16 @@
+select
+  -- Required Columns
+  subnet_id as resource,
+  case
+    when map_public_ip_on_launch = 'false' then 'ok'
```
Comment (khushboo9024) [inline query/vpc/vpc_subnet_auto_assign_public_ip_disabled.sql:5]:
```suggestion
    when map_public_ip_on_launch then 'ok'
```

File: query/vpc/vpc_subnet_auto_assign_public_ip_disabled.sql [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+  end status,
+  case
+    when map_public_ip_on_launch = 'false' then title || ' SUBNET AUTO ASSIGN PUBLIC IP DISABLED.'
+    else title || ' SUBNET AUTO ASSIGN PUBLIC IP ENABLED.'
+  end reason,
+  -- Additional Dimensions
+  region,
+  account_id
+from
+  aws_vpc_subnet
```
Comment (khushboo9024) [inline query/vpc/vpc_subnet_auto_assign_public_ip_disabled.sql:16]:
```suggestion
  aws_vpc_subnet;
```

File: foundational_security/apigateway.sp [hunk 8]
```diff
@@ -66,4 +67,17 @@ control "foundational_security_apigateway_4" {
...
+control "foundational_security_apigateway_5" {
+  title         = "5 API Gateway REST API cache data should be encrypted at rest"
+  description   = "This control checks whether all methods in API Gateway REST API stages that have cache enabled are encrypted. The control fails if any method in an API Gateway REST API stage is configured to cache and the cache is not encrypted."
+  severity      = "medium"
+  sql           = query.apigateway_stage_cache_encryption_at_rest_enabled.sql
+  documentation = file("./foundational_security/docs/foundational_security_apigateway_5.md")
+
+  tags = merge(local.foundational_security_apigateway_common_tags, {
+    foundational_security_item_id  = "apigateway_5"
+    foundational_security_category = "protective_services"
```
Comment (khushboo9024) [inline foundational_security/apigateway.sp:81]:
```suggestion
    foundational_security_category = "encryption_of_data_at_rest'
```
Please check other places as well.

File: foundational_security/ec2.sp [hunk 9]
```diff
@@ -178,4 +193,17 @@ control "foundational_security_ec2_18" {
     foundational_security_item_id  = "ec2_18"
     foundational_security_category = "security_group_configuration"
   })
+}
+
+control "foundational_security_ec2_19" {
+  title         = "19 Security groups should not allow unrestricted access to ports with high risk"
+  description   = "This control checks whether unrestricted incoming traffic for the security groups is accessible to the specified ports that have the highest risk. This control passes when none of the rules in a security group allow ingress traffic from 0.0.0.0/0 for those ports."
+  severity      = "Medium"
```
Comment (khushboo9024) [inline foundational_security/ec2.sp:201]:
```suggestion
  severity      = "medium"
```

File: foundational_security/lambda.sp [hunk 10]
```diff
@@ -38,4 +39,17 @@ control "foundational_security_lambda_2" {
...
     #foundational_security_category = "secure_development"
   })
+}
+
+control "foundational_security_lambda_4" {
+  title         = "4 Lambda functions should have a dead-letter queue configured"
+  description   = "This control checks whether a Lambda function is configured with a dead-letter queue. The control fails if the Lambda function is not configured with a dead-letter queue."
+  severity      = "medium"
+  sql           = query.lambda_function_dead_letter_queue_configured.sql
+  documentation = file("./foundational_security/docs/foundational_security_lambda_2.md")
```
Comment (khushboo9024) [inline foundational_security/lambda.sp:49]:
```suggestion
  documentation = file("./foundational_security/docs/foundational_security_lambda_4.md")
```

File: foundational_security/lambda.sp [hunk 11]
```diff
@@ -38,4 +39,17 @@ control "foundational_security_lambda_2" {
...
+
+  tags = merge(local.foundational_security_lambda_common_tags, {
+    foundational_security_item_id  = "lambda_4"
+    #foundational_security_category = "secure_development"
```
Comment (khushboo9024) [inline foundational_security/lambda.sp:53]:
```suggestion
    foundational_security_category = "logging"
```

File: foundational_security/ec2.sp [hunk 12]
```diff
@@ -141,6 +143,19 @@ control "foundational_security_ec2_10" {
   })
 }
 
+control "foundational_security_ec2_15" {
+  title         = "15 EC2 subnets should not automatically assign public IP addresses"
+  description   = "This control checks whether the assignment of public IPs in Amazon Virtual Private Cloud (Amazon VPC) subnets have MapPublicIpOnLaunch set to FALSE. The control passes if the flag is set to FALSE.."
```
Comment (khushboo9024) [inline foundational_security/ec2.sp:148]:
```suggestion
  description   = "This control checks whether the assignment of public IPs in Amazon Virtual Private Cloud (Amazon VPC) subnets have MapPublicIpOnLaunch set to FALSE. The control passes if the flag is set to FALSE."
```

File: query/rds/rds_db_instance_events_subscription.sql [hunk 13]
```diff
@@ -0,0 +1,18 @@
+select
+  -- Required Columns
+  arn as resource,
+  case
+    when source_type <> 'db-instance' then 'skip'
+    when source_type = 'db-instance' and enabled and event_categories_list @> '["failure", "maintenance", "configuration change"]' then 'ok'
+    else 'alarm'
+  end as status,
+  case
+    when source_type <> 'db-instance' then cust_subscription_id || ' is skipped for critical database instance events.'
```
Comment (rajlearner17) [inline query/rds/rds_db_instance_events_subscription.sql:10]: We should change the reason messaging here - let's not use the word skipped as already the status is skip

`then cust_subscription_id || ' is skipped for critical database instance events.' `can be >>> `then cust_subscription_id || ' event subscription of ' ||  source_type || '.'

File: query/ec2/ec2_classic_lb_connection_draining_enabled.sql [hunk 14]
```diff
@@ -0,0 +1,16 @@
+select
+  -- Required Columns
+  arn as resource,
+  case
+    when connection_draining_enabled then 'ok'
+    else 'alarm'
+  end status,
```
Comment (khushboo9024) [inline query/ec2/ec2_classic_lb_connection_draining_enabled.sql:7]:
```suggestion
  end as status,
```

File: query/ec2/ec2_classic_lb_connection_draining_enabled.sql [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+  case
+    when connection_draining_enabled then 'Classic Load balancer ' || name || ' connection draining is enabled.'
+    else 'Classic Load balancer ' || name || ' connection draining is disabled.'
+  end reason,
```
Comment (khushboo9024) [inline query/ec2/ec2_classic_lb_connection_draining_enabled.sql:11]:
```suggestion
  end as reason,
```

File: query/vpc/vpc_subnet_auto_assign_public_ip_disabled.sql [hunk 16]
```diff
@@ -0,0 +1,16 @@
+select
+  -- Required Columns
+  subnet_id as resource,
+  case
+    when map_public_ip_on_launch = 'false' then 'ok'
+    else 'alarm'
+  end status,
+  case
+    when map_public_ip_on_launch = 'false' then title || ' SUBNET AUTO ASSIGN PUBLIC IP DISABLED.'
```
Comment (khushboo9024) [inline query/vpc/vpc_subnet_auto_assign_public_ip_disabled.sql:9]:
```suggestion
    when map_public_ip_on_launch = 'false' then title || '  auto assign public IP disabled.'
```

File: query/vpc/vpc_subnet_auto_assign_public_ip_disabled.sql [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+    else title || ' SUBNET AUTO ASSIGN PUBLIC IP ENABLED.'
```
Comment (khushboo9024) [inline query/vpc/vpc_subnet_auto_assign_public_ip_disabled.sql:10]:
```suggestion
    else title || ' auto assign public IP enabled.'
```

**Review by khushboo9024 — APPROVED** [review]

**Review by rajlearner17 — APPROVED** [review]
LGTM

**Pull request merged**